File: swift/Sources/SwiftBenchmark/Output.swift
import Foundation

struct SystemInfo: Codable {
    let os: String
    let arch: String
    let cpus: Int
}

struct BenchmarkReport: Codable {
    let language: String
    let timestamp: String
    let swiftVersion: String
    let system: SystemInfo
    let tests: [BenchmarkResult]
    let totalTimeSeconds: Double
    
    enum CodingKeys: String, CodingKey {
        case language
        case timestamp
        case swiftVersion = "swift_version"
        case system
        case tests
        case totalTimeSeconds = "total_time_seconds"
    }
}

class Output {
    static func printResults(_ results: [BenchmarkResult]) {
        print("\n=== BENCHMARK RESULTS ===")
        for result in results {
            print("Test: \(result.test)")
            print("  Duration: \(result.durationNs) ns")
            print("  Memory: \(result.memoryBytes) bytes")
            print("  Operations: \(result.operations)")
            print("  Ops/sec: \(String(format: "%.2f", result.opsPerSec))")
            print()
        }
    }
    
    static func saveResultsToJSON(_ results: [BenchmarkResult], totalTime: Double) throws {
        let report = BenchmarkReport(
            language: "swift",
            timestamp: ISO8601DateFormatter().string(from: Date()),
            swiftVersion: getSwiftVersion(),
            system: SystemInfo(
                os: getOSName(),
                arch: getArchitecture(),
                cpus: ProcessInfo.processInfo.processorCount
            ),
            tests: results,
            totalTimeSeconds: totalTime
        )
        
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        let timestamp = formatter.string(from: Date())
        let filename = "benchmark_swift_\(timestamp).json"
        
        let encoder = JSONEncoder()
        encoder.outputFormatting = .prettyPrinted
        let jsonData = try encoder.encode(report)
        
        let url = URL(fileURLWithPath: filename)
        try jsonData.write(to: url)
        print("Results saved to: \(filename)")
    }
    
    private static func getSwiftVersion() -> String {
        #if swift(>=5.8)
        return "5.8+"
        #elseif swift(>=5.7)
        return "5.7"
        #elseif swift(>=5.6)
        return "5.6"
        #else
        return "5.5 or earlier"
        #endif
    }
    
    private static func getOSName() -> String {
        #if os(macOS)
        return "macOS"
        #elseif os(iOS)
        return "iOS"
        #elseif os(watchOS)
        return "watchOS"
        #elseif os(tvOS)
        return "tvOS"
        #elseif os(Linux)
        return "Linux"
        #else
        return "Unknown"
        #endif
    }
    
    private static func getArchitecture() -> String {
        #if arch(x86_64)
        return "x86_64"
        #elseif arch(arm64)
        return "arm64"
        #elseif arch(arm)
        return "arm"
        #else
        return "unknown"
        #endif
    }
}

File: scala/src/main/scala/com/benchmark/Output.scala
package com.benchmark

import org.json4s._
import org.json4s.native.JsonMethods._
import org.json4s.native.Serialization
import org.json4s.native.Serialization.write
import java.io.{File, PrintWriter}
import java.time.{LocalDateTime, ZoneOffset}
import java.time.format.DateTimeFormatter

case class SystemInfo(
  os: String,
  arch: String,
  cpus: Int
)

case class BenchmarkReport(
  language: String,
  timestamp: String,
  scalaVersion: String,
  system: SystemInfo,
  tests: List[BenchmarkResult],
  totalTimeSeconds: Double
)

object Output {
  implicit val formats: Formats = Serialization.formats(NoTypeHints)

  def printResults(results: List[BenchmarkResult]): Unit = {
    println("\n=== BENCHMARK RESULTS ===")
    results.foreach { result =>
      println(s"Test: ${result.test}")
      println(s"  Duration: ${result.durationNs} ns")
      println(s"  Memory: ${result.memoryBytes} bytes")
      println(s"  Operations: ${result.operations}")
      println(f"  Ops/sec: ${result.opsPerSec}%.2f")
      println()
    }
  }

  def saveResultsToJSON(results: List[BenchmarkResult], totalTime: Double): Unit = {
    val report = BenchmarkReport(
      language = "scala",
      timestamp = LocalDateTime.now(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME) + "Z",
      scalaVersion = getScalaVersion,
      system = SystemInfo(
        os = getOSName,
        arch = getArchitecture,
        cpus = Runtime.getRuntime.availableProcessors()
      ),
      tests = results,
      totalTimeSeconds = totalTime
    )

    val timestamp = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"))
    val filename = s"benchmark_scala_$timestamp.json"

    val jsonString = write(report)
    val prettyJson = pretty(render(parse(jsonString)))

    val writer = new PrintWriter(new File(filename))
    try {
      writer.write(prettyJson)
    } finally {
      writer.close()
    }

    println(s"Results saved to: $filename")
  }

  private def getScalaVersion: String = {
    scala.util.Properties.versionString
  }

  private def getOSName: String = {
    System.getProperty("os.name")
  }

  private def getArchitecture: String = {
    System.getProperty("os.arch")
  }
}